python/ray/_private/runtime_env/physical_mode_execution_utils.py

```diff
@@ -0,0 +1,177 @@
+# 2. Check the existence for cgroup V2.
+# 3. Add error handling; for now just use boolean return to differentiate success or
+# failure. Better to log the error.
+
+from ray._private.runtime_env import physical_mode_context
+
+import sys
+import os
+import shutil
+
+# There're two types of memory cgroup constraints:
+# 1. For those with limit memory capped, they will be created a dedicated cgroup;
+# 2. For those without limit memory specified, they will be added to the default memory
+# cgroup.
+DEFAULT_MEM_CGROUP_UUID = "default_cgroup_memory_uuid"
+
+
+def _setup_new_cgroup_for_memory(
+    physical_exec_ctx: physical_mode_context.PhysicalModeExecutionContext,
+) -> bool:
+    """Setup a new cgroup for memory consumption. Return whether memory constraint is
+    applied successfully."""
+    if not physical_exec_ctx.memory_cgroup_uuid:
+        return False
+
+    if physical_exec_ctx.memory_cgroup_uuid == DEFAULT_MEM_CGROUP_UUID:
+        return False
+
+    cgroup_folder = f"/sys/fs/cgroup/{physical_exec_ctx.memory_cgroup_uuid}"
+    os.makedirs(cgroup_folder, exist_ok=True)
+    pid = os.getpid()
+
+    with open(f"{cgroup_folder}/cgroup.procs", "w") as f:
+        f.write(str(pid))
+
+    if physical_exec_ctx.min_memory:
+        with open(f"{cgroup_folder}/memory.min", "w") as f:
+            f.write(str(physical_exec_ctx.min_memory))
+
+    if physical_exec_ctx.max_memory:
+        with open(f"{cgroup_folder}/memory.max", "w") as f:
+            f.write(str(physical_exec_ctx.max_memory))
+
+    return True
+
+
+def _update_default_cgroup_for_memory_at_setup(
+    physical_exec_ctx: physical_mode_context.PhysicalModeExecutionContext,
+) -> bool:
+    """Add current process to the default cgroup, and update reserved memory if
+    necessary."""
+    if physical_exec_ctx.memory_cgroup_uuid != DEFAULT_MEM_CGROUP_UUID:
+        return False
+
+    # Sanity check.
+    assert physical_exec_ctx.max_memory is None
+
+    cgroup_folder = f"/sys/fs/cgroup/{DEFAULT_MEM_CGROUP_UUID}"
+    os.makedirs(cgroup_folder, exist_ok=True)
+    pid = os.getpid()
+
+    cgroup_procs_path = f"{cgroup_folder}/cgroup.procs"
+    with open(cgroup_procs_path, "a") as f:
+        f.write(f"{pid}\n")
+
+    if not physical_exec_ctx.min_memory:
+        return True
+
+    cur_memory_min = 0
+    if os.path.exists(f"{cgroup_folder}/memory.min"):
+        with open(f"{cgroup_folder}/memory.min", "r") as f:
+            cur_memory_min = int(f.read().strip())
+
+    new_memory_min = cur_memory_min + physical_exec_ctx.min_memory
+    with open(f"{cgroup_folder}/memory.min", "w") as f:
+        f.write(str(new_memory_min))
+
+    return True
+
+
+def _setup_cgroup_for_memory(
+    physical_exec_ctx: physical_mode_context.PhysicalModeExecutionContext,
+) -> bool:
+    """Setup cgroup for memory consumption. Return whether memory constraint is applied
+    successfully."""
+    # Create a new dedicated cgroup if max memory specified.
+    if physical_exec_ctx.max_memory:
```

Comment:
I think if any of the max is set, you will create a new cgroup.

Reply:
Yes exactly, a new cgroup (with a new UUID provided) will be created; for the util function here we simply set the parameters for the cgroup.